### PR TITLE
Added idle connection timeout to HTTPProxy

### LINKED ISSUE #3313: Upstream Idle Connection Timeout Setting
**Need to understand, is there any configurable upstream idle connection setting at contour:**

As per envoy proxy timeout settings [envoy timeouts](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts#connection-timeouts) , upstream idle connections  use the common_http_protocol_options field in the Cluster’s extension_protocol_options, keyed by envoy.extensions.upstreams.http.v3.HttpProtocolOptions,  didn't find any corresponding option at contour to override via httpproxy or contour configuration 

We have services running on tomcat application server in kubernetes and idle connections being closed by tomcat with  keepalive setting 20 sec but still envoy using those connections to send request resulting upstream_cx_destroy_remote_with_active_rq 503 error and tomcat is closing connections before envoy 

Route level timeoutPolicy idle timeout default 5m only dealing with active stream but not idle connections, we have tried with route level idle timeout to match tomcat keep alive settings but it is not working  and still there are upstream_cx_destroy_remote_with_active_rq 503 errors as  tomcat is closing connections before envoy 

Flow:  Client --> Envoy(Kubernetes)-->Tomact App(Kubernetes)  503 with upstream to envoy 

**We need to have some configurable setting for upstream idle connections via contour configuration or httpproxy, and please provide details on what is the envoy default upstream idle connection timeout value  and how we can override it from contour/httpproxy**

### PR DESCRIPTION
**Pull request by tsaarni**
This change adds field `HTTPProxy.spec.routes.timeoutPolicy.idleConnection`. The field sets timeout for how long the upstream connection will be kept idle between requests before disconnecting it. This [connection timeout](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts#connection-timeouts) is different than previously existing field `timeoutPolicy.idle`, which controls the [stream timeout](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts#stream-timeouts): a timeout for single request-response. Altough connection timeout might be what user wanted in the first place, the existing field and functionality is kept for backwards compatibility and new field was introduced for the connection timeout.

When `timeoutPolicy.idleConnection` is not set then Envoy defaults to [1 hour](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-httpprotocoloptions). User can also set `infinite` to disable the timeout

Timeout is set in Envoy `Cluster` by using [`HttpProtocolOptions`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/upstreams/http/v3/http_protocol_options.proto.html). If two `HTTPProxies` refer to single upstream service and are otherwise equal except for the `timeoutpolicy.idleConnection` value, then separate `Clusters` will be created for both to allow different timeouts.

Fixes #3313
    
Signed-off-by: Tero Saarni <tero.saarni@est.tech>

### CODE REVIEW

**Review by youngnick — APPROVED** [review]
LGTM, nice change.

**Comment by tsaarni** [comment]
Since this is `HTTPProxy` API change, I'd appreciate another pair of eyes on the PR :smiley:  
Will merge after that.

File: internal/dag/policy.go
Comment (sunjayBhatia) [inline internal/dag/policy.go:409]: missing `IdleConnectionTimeout` here?

File: internal/dag/dag.go
Comment (sunjayBhatia) [inline internal/dag/dag.go:874]: ExtensionCluster already has a TimeoutPolicy field, could we collapse this and other timeout fields into there? just did a quick look, not sure if it totally makes sense etc.

**Review by sunjayBhatia — COMMENTED** [review]
just a couple small things, otherwise looks good, nice addition + naming/description of the field

File: internal/dag/policy_test.go
Comment (skriss) [inline internal/dag/policy_test.go:235]: would be nice to add cases for invalid & valid idle connection timeouts here, for consistency

**Review by skriss — COMMENTED** [review]
Looks good 👍 , one inline comment on test coverage; I also noticed there's a `timeoutpolicy_test.go` in featuretests, would be nice to extend that to cover this new field for completeness.

File: internal/dag/dag.go
Comment (tsaarni) [reply internal/dag/dag.go]: There was small "side-effect" that maybe makes this change bit illogical:  

`TimeoutPolicy` was previously only for `Route` but now after repurposing the same struct for `Cluster` as well, we will have complete set of timeout values included in both `dag.Route` and `dag.Cluster`. To my surprise, two tests in `builder_test.go` started failing:  the tests were about changing response timeout, which previously meant `Route.TimeoutPolicy.ResponseTimeout` was expected to be set. Now after this change, the test must also expect `Cluster.TimeoutPolicy.ResponseTimeout` to be set, even if response timeout makes no sense in context of `dag.Cluster`.

I'm OK with this small illogicality, but I made the change as separate commit so I can easily revert it as well, if wanted!

File: internal/dag/policy.go
Comment (tsaarni) [reply internal/dag/policy.go]: Now added.

File: internal/dag/policy_test.go
Comment (tsaarni) [reply internal/dag/policy_test.go:235]: Added tests for invalid, valid timeout and infinite.

**Comment by tsaarni** [comment]
Thank you for review, good comments!

> Looks good 👍 , one inline comment on test coverage; I also noticed there's a `timeoutpolicy_test.go` in featuretests, would be nice to extend that to cover this new field for completeness.

I've now added test case to featuretests as well,

File: internal/dag/dag.go
Comment (sunjayBhatia) [reply internal/dag/dag.go]: hm yeah maybe not worth the confusion, thanks for looking into that, ok if you would rather keep it as-is and revert

Comment (skriss) [reply internal/dag/dag.go]: Yeah I'm fine backing it out too, but can live with either way.

Comment (youngnick) [reply internal/dag/dag.go]: It doesn't seem like any of these are good options, so I guess we're choosing the least worst?

- Keep what we have here (timeoutpolicy the same, ResponseTimeout must be set for tests)
- revert what we have here, what happens?

Is there a better way to do this? Should we have two timeoutpolicy structs that are mostly the same except for this?

Comment (tsaarni) [reply internal/dag/dag.go]: > Is there a better way to do this? Should we have two timeoutpolicy structs that are mostly the same except for this?

Exactly @youngnick, I got stalled because I started to think that would be the preferred way: specific timeout struct for route, another for cluster. I then got distracted by other work, so sorry for delay. I should try arrange some time soon! :)

Comment (youngnick) [reply internal/dag/dag.go]: Awesome, I'll keep un-staling this then :)

Comment (tsaarni) [reply internal/dag/dag.go]: Splitting `TimeoutPolicy` into `RouteTimeoutPolicy` and `ClusterTimeoutPolicy` worked OK, but I noticed that `dag.ExtensionCluster` behaves bit like a mixture of route and cluster: in addition of creating a cluster it is also used to configure `ExtAuthz_GrpcService` which  has response timeout just like `Route` does. So to follow the DRY principle, I ended up with following:

* `Route` has timeout policy attribute of type `RouteTimeoutPolicy`
* `Cluster` has `ClusterTimeoutPolicy`
* `ExtensionCluster` has both `RouteTimeoutPolicy` and `ClusterTimeoutPolicy`

and now `Route` does not contain timeout parameters that belong to `Cluster` and the other way around - with the negative side effect that `ExtensionCluster` needed both, but what do you think?

File: internal/dag/builder_test.go
Comment (skriss) [inline internal/dag/builder_test.go:11393]: guessing this is left over from debugging :)

File: internal/dag/extension_processor.go
Comment (skriss) [inline internal/dag/extension_processor.go:87]: should we be getting and using the cluster timeout policy here too? I don't see anywhere where it's being set otherwise

**Review by skriss — COMMENTED** [review]
I think this breakout makes sense and is helpful for organizing things. One question on the ExtensionCluster impl.

File: internal/dag/builder_test.go
Comment (tsaarni) [reply internal/dag/builder_test.go]: Sorry, my bad! Glad you spotted it, now fixed 😊

File: internal/dag/dag.go
Comment (skriss) [inline internal/dag/dag.go:864]: oops 😄

File: internal/dag/extension_processor.go
Comment (skriss) [reply internal/dag/extension_processor.go]: 😄  no problem, we've all done it. Updates look good!

**Review by skriss — APPROVED** [review]
LGTM, just one tiny godoc nit above

File: internal/dag/dag.go
Comment (tsaarni) [reply internal/dag/dag.go]: Uhh :) I've now removed that unfinished sentense.  

I was originally going to point out there in comment that the response timeout can come from that struct (originating from `ExtensionService.spec.timeoutPolicy.response`) or it may be overwritten by value from `HTTPProxy.spec.virtualhost.authorization.responseTimeout`. This can be seen in [code here](https://github.com/projectcontour/contour/blob/2b7e6ccdc9fdcce787bab0e0bf3e753fd62a41b9/internal/dag/httpproxy_processor.go#L302-L306) and it was existing functionality but it confused me a bit - and without this logic, which I think is not described in API docs, the `RouteTimeoutPolicy` could be even removed from `ExtensionCluster`.

**Comment by sivakrishna-git** [comment]
@tsaarni : Please update documentation https://projectcontour.io/docs/main/config/request-routing/#response-timeouts as well

File: internal/dag/policy.go
Comment (ravilr) [inline internal/dag/policy.go:433]: @tsaarni @sunjayBhatia @skriss while upgrade testing from contour-v1.19 to latest v1.22, seeing this issue where ContourConfig file has timeout.connect-timeout set to 5s, but the final envoy config rendered by the xdscache keeps fluctuating certain Cluster's connect_timeout b/w 5s and 2s on any onChanges(). it is expected to be 5s for all dynamic Cluster's connect_timeout.

I think this resetting of Cluster's ConnectTimeout=0 seems to be causing this. Why is this resetting Cluster's ConnectTimeout when a Route's TimeoutPolicy isn't set ? shouldn't this be setting Cluster's ConnectTimeout to passed func arg connectTimeout also ? (just like in the final return below).

Comment (tsaarni) [reply internal/dag/policy.go:433]: @ravilr You are correct, thanks for spotting this!

**Pull request merged**
